File: client/src/utils/practiceCalendarLayout.ts
/**
 * 训练打卡月历布局（Q-B3 / v1.1.0）
 *
 * 纯函数；输入 month=YYYY-MM 与 practice_logs 聚合 count，输出 6×7 网格。
 */

export interface PracticeCalendarCell {
  /** YYYY-MM-DD；padding 格为空串 */
  dateKey: string
  day: number
  count: number
  isToday: boolean
  inMonth: boolean
}

export interface PracticeCalendarGrid {
  monthKey: string
  /** 固定 6 行 × 7 列 */
  weeks: PracticeCalendarCell[][]
  monthTotal: number
}

const WEEKDAY_LABELS = ['一', '二', '三', '四', '五', '六', '日'] as const

export function practiceCalendarWeekdayLabels(): readonly string[] {
  return WEEKDAY_LABELS
}

/** 解析 YYYY-MM → { year, month }；非法返回 null */
export function parseMonthKey(monthKey: string): { year: number; month: number } | null {
  const m = monthKey.match(/^(\d{4})-(\d{2})$/)
  if (!m) return null
  const year = Number(m[1])
  const month = Number(m[2])
  if (!year || month < 1 || month > 12) return null
  return { year, month }
}

/** 本地 today → YYYY-MM-DD */
export function localDateKey(d: Date = new Date()): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

export function monthKeyNow(d: Date = new Date()): string {
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`
}

/** 上/下月 monthKey */
export function shiftMonthKey(monthKey: string, delta: number): string {
  const parsed = parseMonthKey(monthKey)
  if (!parsed) return monthKeyNow()
  let { year, month } = parsed
  month += delta
  while (month < 1) {
    month += 12
    year -= 1
  }
  while (month > 12) {
    month -= 12
    year += 1
  }
  return `${year}-${String(month).padStart(2, '0')}`
}

/** 从 practice_logs 列表聚合 dateKey → count */
export function aggregatePracticeCounts(
  logs: { practice_date?: string | null }[],
): Map<string, number> {
  const counts = new Map<string, number>()
  for (const log of logs) {
    const key = (log.practice_date ?? '').slice(0, 10)
    if (!key) continue
    counts.set(key, (counts.get(key) ?? 0) + 1)
  }
  return counts
}

export function buildPracticeCalendarGrid(
  monthKey: string,
  counts: Map<string, number>,
  todayKey: string = localDateKey(),
): PracticeCalendarGrid {
  const parsed = parseMonthKey(monthKey)
  if (!parsed) {
    return { monthKey, weeks: [], monthTotal: 0 }
  }
  const { year, month } = parsed
  const first = new Date(year, month - 1, 1)
  const lastDay = new Date(year, month, 0).getDate()
  // 周一为 0（ISO 风格）
  const firstWeekday = (first.getDay() + 6) % 7

  let monthTotal = 0
  for (let day = 1; day <= lastDay; day += 1) {
    const dd = String(day).padStart(2, '0')
    const mm = String(month).padStart(2, '0')
    monthTotal += counts.get(`${year}-${mm}-${dd}`) ?? 0
  }

  const cells: PracticeCalendarCell[] = []
  for (let i = 0; i < firstWeekday; i += 1) {
    cells.push({ dateKey: '', day: 0, count: 0, isToday: false, inMonth: false })
  }
  for (let day = 1; day <= lastDay; day += 1) {
    const dd = String(day).padStart(2, '0')
    const mm = String(month).padStart(2, '0')
    const dateKey = `${year}-${mm}-${dd}`
    cells.push({
      dateKey,
      day,
      count: counts.get(dateKey) ?? 0,
      isToday: dateKey === todayKey,
      inMonth: true,
    })
  }
  while (cells.length % 7 !== 0) {
    cells.push({ dateKey: '', day: 0, count: 0, isToday: false, inMonth: false })
  }
  while (cells.length < 42) {
    cells.push({ dateKey: '', day: 0, count: 0, isToday: false, inMonth: false })
  }

  const weeks: PracticeCalendarCell[][] = []
  for (let i = 0; i < cells.length; i += 7) {
    weeks.push(cells.slice(i, i + 7))
  }
  return { monthKey, weeks, monthTotal }
}

export function formatMonthTitle(monthKey: string): string {
  const parsed = parseMonthKey(monthKey)
  if (!parsed) return monthKey
  return `${parsed.year}年${parsed.month}月`
}

File: app/lib/core/practice_calendar_layout.dart
import '../data/models/training.dart';

/// 训练打卡月历布局：对照 client/src/utils/practiceCalendarLayout.ts。

class PracticeCalendarCell {
  final String dateKey;
  final int day;
  final int count;
  final bool isToday;
  final bool inMonth;

  const PracticeCalendarCell({
    required this.dateKey,
    required this.day,
    required this.count,
    required this.isToday,
    required this.inMonth,
  });
}

class PracticeCalendarGrid {
  final String monthKey;
  final List<List<PracticeCalendarCell>> weeks;
  final int monthTotal;

  const PracticeCalendarGrid({
    required this.monthKey,
    required this.weeks,
    required this.monthTotal,
  });
}

const practiceCalendarWeekdayLabels = ['一', '二', '三', '四', '五', '六', '日'];

({int year, int month})? parseMonthKey(String monthKey) {
  final m = RegExp(r'^(\d{4})-(\d{2})$').firstMatch(monthKey);
  if (m == null) return null;
  final year = int.tryParse(m.group(1)!);
  final month = int.tryParse(m.group(2)!);
  if (year == null || month == null || month < 1 || month > 12) return null;
  return (year: year, month: month);
}

String localDateKey([DateTime? d]) {
  final x = d ?? DateTime.now();
  return '${x.year}-${_two(x.month)}-${_two(x.day)}';
}

String monthKeyNow([DateTime? d]) {
  final x = d ?? DateTime.now();
  return '${x.year}-${_two(x.month)}';
}

String shiftMonthKey(String monthKey, int delta) {
  final parsed = parseMonthKey(monthKey);
  if (parsed == null) return monthKeyNow();
  var year = parsed.year;
  var month = parsed.month + delta;
  while (month < 1) {
    month += 12;
    year -= 1;
  }
  while (month > 12) {
    month -= 12;
    year += 1;
  }
  return '$year-${_two(month)}';
}

Map<String, int> aggregatePracticeCounts(Iterable<PracticeLogItem> logs) {
  final counts = <String, int>{};
  for (final log in logs) {
    final raw = log.practiceDate ?? '';
    final key = raw.length >= 10 ? raw.substring(0, 10) : '';
    if (key.isEmpty) continue;
    counts[key] = (counts[key] ?? 0) + 1;
  }
  return counts;
}

PracticeCalendarGrid buildPracticeCalendarGrid(
  String monthKey,
  Map<String, int> counts, [
  String? todayKey,
]) {
  final today = todayKey ?? localDateKey();
  final parsed = parseMonthKey(monthKey);
  if (parsed == null) {
    return PracticeCalendarGrid(monthKey: monthKey, weeks: const [], monthTotal: 0);
  }
  final year = parsed.year;
  final month = parsed.month;
  final first = DateTime(year, month, 1);
  final lastDay = DateTime(year, month + 1, 0).day;
  final firstWeekday = (first.weekday + 6) % 7; // Mon=0

  var monthTotal = 0;
  for (var day = 1; day <= lastDay; day++) {
    final key = '$year-${_two(month)}-${_two(day)}';
    monthTotal += counts[key] ?? 0;
  }

  final cells = <PracticeCalendarCell>[];
  for (var i = 0; i < firstWeekday; i++) {
    cells.add(const PracticeCalendarCell(
        dateKey: '', day: 0, count: 0, isToday: false, inMonth: false));
  }
  for (var day = 1; day <= lastDay; day++) {
    final key = '$year-${_two(month)}-${_two(day)}';
    cells.add(PracticeCalendarCell(
      dateKey: key,
      day: day,
      count: counts[key] ?? 0,
      isToday: key == today,
      inMonth: true,
    ));
  }
  while (cells.length % 7 != 0) {
    cells.add(const PracticeCalendarCell(
        dateKey: '', day: 0, count: 0, isToday: false, inMonth: false));
  }
  while (cells.length < 42) {
    cells.add(const PracticeCalendarCell(
        dateKey: '', day: 0, count: 0, isToday: false, inMonth: false));
  }

  final weeks = <List<PracticeCalendarCell>>[];
  for (var i = 0; i < cells.length; i += 7) {
    weeks.add(cells.sublist(i, i + 7));
  }
  return PracticeCalendarGrid(
      monthKey: monthKey, weeks: weeks, monthTotal: monthTotal);
}

String formatMonthTitle(String monthKey) {
  final parsed = parseMonthKey(monthKey);
  if (parsed == null) return monthKey;
  return '${parsed.year}年${parsed.month}月';
}

String _two(int n) => n.toString().padLeft(2, '0');
